### PR TITLE
fix(CodeTransformation): refresh VFS if diff.patch not found

### PR DESCRIPTION
**Pull request by dhasani23**
## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Description

We've had some internal users as well as customer issues where the `diff.patch` file is successfully downloaded, but sometimes cannot be found. This is resolved by refreshing IntelliJ's virtual file system -- once the file is downloaded, we need to refresh to ensure the file is recognized / seen.

I also added a "submit feedback" hyperlink to the notifications we show once a job fails.

## Checklist
- [X] My code follows the code style of this project
- [ ] I have added tests to cover my changes
- [ ] A short description of the change has been added to the **[CHANGELOG](https://github.com/aws/aws-toolkit-jetbrains/blob/master/CONTRIBUTING.md#contributing-via-pull-requests)** if the change is customer-facing in the IDE.
- [ ] I have added metrics for my changes (if required)
 
## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerSession.kt [hunk 1]
```diff
@@ -284,6 +284,7 @@ class CodeModernizerSession(
         if (!shouldStop.get()) {
             CodetransformTelemetry.logApiLatency(
                 codeTransformApiNames = CodeTransformApiNames.UploadZip,
+                codeTransformUploadId = createUploadUrlResponse.uploadId(),
```
Comment (dhasani23) [inline plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerSession.kt:287]: This was missing for some reason, adding it now

File: plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeTransformUtils.kt [hunk 2]
```diff
@@ -196,7 +196,7 @@ suspend fun JobId.pollTransformationStatusAndPlan(
                     CodetransformTelemetry.jobStatusChanged(
                         codeTransformSessionId = CodeTransformTelemetryState.instance.getSessionId(),
                         codeTransformJobId = this.id,
-                        codeTransformStatus = newStatus.toString()
+                        codeTransformStatus = "$state -> $newStatus", // old status -> new status
```
Comment (dhasani23) [inline plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeTransformUtils.kt:199]: Useful so that we can track the old state and new state that a job transitioned to.

File: plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/model/CodeModernizerArtifact.kt [hunk 3]
```diff
@@ -99,6 +101,14 @@ open class CodeModernizerArtifact(
...
+            val patch = patchesDir.walk().map { fileSystem.findFileByNioFile(it) }.toList()
+            if (patch[0] == null) {
+                ApplicationManager.getApplication().invokeLater {
+                    ApplicationManager.getApplication().runWriteAction {
+                        VirtualFileManager.getInstance().syncRefresh() // try refreshing Virtual File System if diff.patch not found
+                    }
+                }
+            }
             return patchesDir.walk()
                 .map { fileSystem.findFileByNioFile(it) ?: throw RuntimeException("Could not find patch") }
```
Comment (dhasani23) [inline plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/model/CodeModernizerArtifact.kt:113]: This exception was being thrown repeatedly, as seen by internal and external users and our dashboard.

File: plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/model/CodeModernizerArtifact.kt [hunk 4]
```diff
@@ -99,6 +101,14 @@ open class CodeModernizerArtifact(
...
             }
+            val patch = patchesDir.walk().map { fileSystem.findFileByNioFile(it) }.toList()
+            if (patch[0] == null) {
+                ApplicationManager.getApplication().invokeLater {
+                    ApplicationManager.getApplication().runWriteAction {
+                        VirtualFileManager.getInstance().syncRefresh() // try refreshing Virtual File System if diff.patch not found
+                    }
+                }
+            }
             return patchesDir.walk()
```
Comment (mk-fan) [inline plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/model/CodeModernizerArtifact.kt:112]: Question: Do we need to do `patchesDir.walk()` again if the first call resulted in success?

Comment (dhasani23) [reply plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/model/CodeModernizerArtifact.kt:112]: Perhaps not but there's only 1 file (the `diff.patch`) in that directory so IMO this is an almost negligible optimization.

File: plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/model/CodeModernizerArtifact.kt [hunk 5]
```diff
@@ -99,6 +101,14 @@ open class CodeModernizerArtifact(
             if (!patchesDir.isDirectory()) {
                 throw RuntimeException("Expected root for patches was not a directory.")
             }
+            val patch = patchesDir.walk().map { fileSystem.findFileByNioFile(it) }.toList()
+            if (patch[0] == null) {
+                ApplicationManager.getApplication().invokeLater {
```
Comment (mk-fan) [inline plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/model/CodeModernizerArtifact.kt:106]: Will this guarantee `syncRefresh` to run before the next `patchesDir.walk()` call? Not super familiar with invokeLater

File: plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/model/CodeModernizerArtifact.kt (hunk 4 above)
Comment (dhasani23) [reply plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/model/CodeModernizerArtifact.kt:112]: Will fix this anyway, doesn't hurt

File: plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeTransformUtils.kt (hunk 2 above)
Comment (damntrecky) [reply plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeTransformUtils.kt:199]: Blocker: I was expecting a new field. This is not backwards compatible with the current `codeTransformStatus` emission. With this change we would also have to parse this field to get the status. It will be more difficult to visually graph this.

Comment (damntrecky) [reply plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeTransformUtils.kt:199]: Also this change would need to be reflected in VSCode to have telemetry parity.

File: plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/model/CodeModernizerArtifact.kt (hunk 5 above)
Comment (damntrecky) [reply plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/model/CodeModernizerArtifact.kt:106]: Question: I'm not sure I understand the problem here. Calling `invokeLater` with the sync needs to happen before the `return patchesDir.walk()` statement correct?

This `invokeLater` function is async so the the `return patchesDir.walk()` will possibly get called before the `invokeLater` finishes.

File: plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeTransformUtils.kt (hunk 2 above)
Comment (dhasani23) [reply plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeTransformUtils.kt:199]: Makes sense. Will add a new field and yes this is planned for VS Code too.

File: plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/model/CodeModernizerArtifact.kt (hunk 5 above)
Comment (damntrecky) [reply plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/model/CodeModernizerArtifact.kt:106]: Check this example output:
```
    println("Main thread: Before invokeLater")

    SwingUtilities.invokeLater {
        println("Running on EDT")
        Thread.sleep(1500) // Simulate a long-running task
        println("Done running on EDT")
    }


    println("Main thread: After invokeLater")
    Thread.sleep(2000) // Wait for 1 second
    println("Main thread: Exiting")
```

Outputs the following:

```
Main thread: Before invokeLater
Main thread: After invokeLater
Running on EDT
Done running on EDT
Main thread: Exiting
```

Comment (damntrecky) [reply plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/model/CodeModernizerArtifact.kt:106]: If you want to test something out quickly. Try the online [kotlin playground](https://play.kotlinlang.org/#eyJ2ZXJzaW9uIjoiMS45LjIyIiwicGxhdGZvcm0iOiJqYXZhIiwiYXJncyI6IiIsIm5vbmVNYXJrZXJzIjp0cnVlLCJ0aGVtZSI6ImlkZWEiLCJjb2RlIjoiaW1wb3J0IGphdmF4LnN3aW5nLlN3aW5nVXRpbGl0aWVzXG4vKipcbiAqIFlvdSBjYW4gZWRpdCwgcnVuLCBhbmQgc2hhcmUgdGhpcyBjb2RlLlxuICogcGxheS5rb3RsaW5sYW5nLm9yZ1xuICovXG5mdW4gbWFpbigpIHtcblxuICAgIHByaW50bG4oXCJNYWluIHRocmVhZDogQmVmb3JlIGludm9rZUxhdGVyXCIpXG5cblxuICAgIFN3aW5nVXRpbGl0aWVzLmludm9rZUxhdGVyIHtcblxuICAgICAgICBwcmludGxuKFwiUnVubmluZyBvbiBFRFRcIilcblxuICAgICAgICBUaHJlYWQuc2xlZXAoMTUwMCkgLy8gU2ltdWxhdGUgYSBsb25nLXJ1bm5pbmcgdGFza1xuXG4gICAgICAgIHByaW50bG4oXCJEb25lIHJ1bm5pbmcgb24gRURUXCIpXG5cbiAgICB9XG5cblxuICAgIHByaW50bG4oXCJNYWluIHRocmVhZDogQWZ0ZXIgaW52b2tlTGF0ZXJcIilcblxuICAgIFRocmVhZC5zbGVlcCgyMDAwKSAvLyBXYWl0IGZvciAxIHNlY29uZFxuXG5cbiAgICBwcmludGxuKFwiTWFpbiB0aHJlYWQ6IEV4aXRpbmdcIilcbn0ifQ==)

Comment (dhasani23) [reply plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/model/CodeModernizerArtifact.kt:106]: That's a good point. We also have `invokeAndWait` which appears to work fine too. Though it seems like you're not supposed to call that from the EDT as it will lead to deadlock and throw an Exception. This didn't happen for me but want to learn more about this before doing it. Also want @arangatang to give his thoughts.

Comment (rli) [reply plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/model/CodeModernizerArtifact.kt:106]: what about `refreshAndFindFileByPath`? no need to refresh the entire VFS

File: plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/model/CodeModernizerSessionContext.kt [hunk 6]
```diff
@@ -66,10 +66,6 @@ data class CodeModernizerSessionContext(
         return isIdea
     }
 
-    /**
-     * Can eventually modify to use an approach based on walkTopDown instead of VfsUtil.collectChildrenRecursively(root) in createZipWithModuleFiles.
-     * We now recurse the file tree twice and then filter which hurts performance for large projects.
-     */
```
Comment (dhasani23) [inline plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/model/CodeModernizerSessionContext.kt:72]: Removed this and instead created a SIM for it

File: plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/model/CodeModernizerArtifact.kt [hunk 7]
```diff
@@ -99,7 +100,17 @@ open class CodeModernizerArtifact(
             if (!patchesDir.isDirectory()) {
                 throw RuntimeException("Expected root for patches was not a directory.")
             }
-            return patchesDir.walk()
+            val patchPathSequence = patchesDir.walk()
+            val patchFileList = patchPathSequence.map { fileSystem.findFileByNioFile(it) }.toList()
+            if (patchFileList[0] == null) {
+                ApplicationManager.getApplication().invokeAndWait {
+                    ApplicationManager.getApplication().runWriteAction {
```
Comment (dhasani23) [inline plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/model/CodeModernizerArtifact.kt:107]: Would like feedback if anyone has a better understanding of the necessity of `invokeAndWait` plus `runWriteAction`. Seems like it comes down to thread safety

Comment (dhasani23) [reply plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/model/CodeModernizerArtifact.kt:107]: This seems to summarize my concerns: https://javarevisited.blogspot.com/2011/09/invokeandwait-invokelater-swing-example.html#axzz8Tjdw80jd

**Review by arangatang — DISMISSED** [review]

File: plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/model/CodeModernizerArtifact.kt [hunk 8]
```diff
@@ -99,7 +100,17 @@ open class CodeModernizerArtifact(
...
+                        // try refreshing Virtual File System if diff.patch not found
+                        VirtualFileManager.getInstance().refreshAndFindFileByNioPath(patchPathSequence.first())
+                    }
```
Comment (arangatang) [inline plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/model/CodeModernizerArtifact.kt:110]: Could we not return this directly? and just call this?

File: plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/model/CodeModernizerArtifact.kt [hunk 9]
```diff
@@ -99,7 +100,17 @@ open class CodeModernizerArtifact(
...
+                }
+            }
+            return patchPathSequence
                 .map { fileSystem.findFileByNioFile(it) ?: throw RuntimeException("Could not find patch") }
```
Comment (damntrecky) [inline plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/model/CodeModernizerArtifact.kt:114]: Leos advice: replace line 114 `map` with `refreshAndFindFileByNioPath` instead of `findFileByNioFile`

Comment (damntrecky) [reply plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/model/CodeModernizerArtifact.kt:114]: Leos comment: We also only return a single patch file, if the `walk()` complicates things, just check array index [0] directly

File: plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt [hunk 10]
```diff
@@ -688,9 +698,6 @@ class CodeModernizerManager(private val project: Project) : PersistentStateCompo
         addCodeModernizeUI(true)
         val maybeUnknownReason = reason ?: message("codemodernizer.notification.warn.invalid_project.description.reason.unknown")
         codeModernizerBottomWindowPanelManager.setProjectInvalidUI(maybeUnknownReason)
-        ApplicationManager.getApplication().invokeLater {
-            runInEdt { ValidationErrorDialog.create(maybeUnknownReason) }
```
Comment (dhasani23) [inline plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt:692]: This was repetitive -- we already have a notification showing the exact same info and it pops up at the same time.

File: plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/model/CodeModernizerArtifact.kt [hunk 11]
```diff
@@ -100,7 +99,7 @@ open class CodeModernizerArtifact(
                 throw RuntimeException("Expected root for patches was not a directory.")
             }
             return patchesDir.walk()
-                .map { fileSystem.findFileByNioFile(it) ?: throw RuntimeException("Could not find patch") }
+                .map { fileSystem.refreshAndFindFileByNioFile(it) ?: throw RuntimeException("Could not find patch") }
```
Comment (dhasani23) [inline plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/model/CodeModernizerArtifact.kt:102]: Zongren tested and it worked

File: plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt [hunk 12]
```diff
@@ -554,6 +567,11 @@ class CodeModernizerManager(private val project: Project) : PersistentStateCompo
             artifactHandler.showTransformationSummary(jobId)
         }
 
+    private fun displayFeedbackNotificationAction() =
+        NotificationAction.createSimple(message("codemodernizer.notification.warn.submit_feedback")) {
+            FeedbackDialog(project, productName = "Amazon Q").showAndGet()
```
Comment (arangatang) [inline plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt:572]: Is there a Constant we can use for `Amazon Q` that we can use instead?

File: plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt [hunk 13]
```diff
@@ -297,6 +297,7 @@ class CodeModernizerManager(private val project: Project) : PersistentStateCompo
             message("codemodernizer.notification.info.transformation_stop.title"),
             message("codemodernizer.notification.info.transformation_stop.content"),
             project,
+            listOf(displayFeedbackNotificationAction())
```
Comment (arangatang) [inline plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt:300]: This is an unrelated change to the VFS refresh. Probably this should have been added in a separate PR.

File: plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt (hunk 10 above)
Comment (arangatang) [reply plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt:692]: Can you add screenshots to the PR with these changes to clearly show the changes (before and after). Makes buy in easier.

File: plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt [hunk 14]
```diff
@@ -578,17 +587,18 @@ class CodeModernizerManager(private val project: Project) : PersistentStateCompo
                 false,
             )
 
-            is CodeModernizerJobCompletedResult.JobFailedInitialBuild -> {
-                ApplicationManager.getApplication().invokeLater {
-                    runInEdt { BuildErrorDialog.create(result.failureReason) }
```
[the remark this answers is not in the document]
Comment (arangatang) [reply plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt:583]: How does this look? The content here is coming from the backend and may be verbose, thats why this one is a dialog instead of a notification as notifications truncates  text when generating the preview. Can you add some screenshots of this and is this dicussed with UX?

**Review by Quoc8234065 — APPROVED** [review]

File: plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt (hunk 12 above)
Comment (damntrecky) [reply plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt:572]: Nice to have: Make this a constant to be shared for cleaner code. Something like `AMAZON_Q_FEEDBACK_DIALOG_KEY`

File: plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt (hunk 14 above)
[the remark this answers is not in the document]
Comment (damntrecky) [reply plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt:583]: +1 For before an after screenshots to help understand the full scope/context.

File: plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/model/CodeModernizerArtifact.kt (hunk 11 above)
Comment (damntrecky) [reply plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/model/CodeModernizerArtifact.kt:102]: Note: This is the main fix for "refresh VFS if diff.patch not found".

**Review by damntrecky — CHANGES_REQUESTED** [review]
Please leave a screen recording and/or screenshots of before and after of the messaging changes.

If you want this PR to go out immediately, I would cherry pick out the VFS one line change into another PR.

File: plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt (hunk 14 above)
[the remark this answers is not in the document]
Comment (dhasani23) [reply plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt:583]: Already discussed with UX (Allie) that we shouldn't be showing a dialog and a notification with the same content, which was happening so far.
image

[the remark this answers is not in the document]
Comment (dhasani23) [reply plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt:583]: The specific text is going to be reviewed & modified later. We previously had this notification + a modal in the middle showing the same exact info. Allie said we only need the notification. Following this logic, I removed the build error dialog / modal too and instead put the same info in a notification.

File: plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt (hunk 10 above)
Comment (dhasani23) [reply plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt:692]: See above

File: plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt (hunk 14 above)
[the remark this answers is not in the document]
Comment (dhasani23) [reply plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt:583]: Screenshot 2024-03-13 at 10 46 02 AM

[the remark this answers is not in the document]
Comment (dhasani23) [reply plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt:583]: before ^

[the remark this answers is not in the document]
Comment (dhasani23) [reply plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerManager.kt:583]: It's the exact same for the pre-build error.

**Comment by dhasani23** [comment]
@damntrecky added screenshots

File: plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeTransformUtils.kt (hunk 2 above)
Comment (dhasani23) [reply plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeTransformUtils.kt]: This will be done in a separate PR

**Review by damntrecky — APPROVED** [review]

**Review by rli — APPROVED** [review]
